admittance_controller/include/admittance_controller/admittance_rule.hpp

```diff
@@ -0,0 +1,215 @@
+  Eigen::VectorXd joint_pos;
+  Eigen::VectorXd joint_vel;
+  Eigen::VectorXd joint_acc;
+  Eigen::Matrix<double, 6, 1> damping;
+  Eigen::Matrix<double, 6, 1> mass;
+  Eigen::Matrix<double, 6, 1> mass_inv;
+  Eigen::Matrix<double, 6, 1> selected_axes;
+  Eigen::Matrix<double, 6, 1> stiffness;
+  Eigen::Matrix<double, 6, 1> wrench_base;
+  Eigen::Matrix<double, 6, 1> admittance_acceleration;
+  Eigen::Matrix<double, 6, 1> admittance_velocity;
+  Eigen::Isometry3d admittance_position;
+  Eigen::Matrix<double, 3, 3> rot_base_control;
+  Eigen::Isometry3d ref_trans_base_ft;
+  std::string ft_sensor_frame;
+};
+
+class AdmittanceRule
+{
+public:
+  explicit AdmittanceRule(
+    const std::shared_ptr<admittance_controller::ParamListener> & parameter_handler)
+  {
+    parameter_handler_ = parameter_handler;
+    parameters_ = parameter_handler_->get_params();
+    num_joints_ = parameters_.joints.size();
+    admittance_state_ = AdmittanceState(num_joints_);
+    reset(num_joints_);
+  }
+
+  /// Configure admittance rule memory using number of joints.
+  controller_interface::return_type configure(
+    const std::shared_ptr<rclcpp_lifecycle::LifecycleNode> & node, const size_t num_joint);
+
+  /// Reset all values back to default
+  controller_interface::return_type reset(const size_t num_joints);
+
+  /**
+   * Calculate all transforms needed for admittance control using the loader kinematics plugin. If the transform does
+   * not exist in the kinematics model, then TF will be used for lookup. The return value is true if all transformation
+   * are calculated without an error
+   * \param[in] current_joint_state
+   * \param[in] reference_joint_state
+   * \param[out] success
+   */
+  bool get_all_transforms(
+    const trajectory_msgs::msg::JointTrajectoryPoint & current_joint_state,
+    const trajectory_msgs::msg::JointTrajectoryPoint & reference_joint_state);
+
+  /**
+   * Updates parameter_ struct if any parameters have changed since last update. Parameter dependent Eigen field
+   * members (end_effector_weight_, cog_pos_, mass_, mass_inv_ stiffness, selected_axes, damping_) are also updated
+   */
+  void apply_parameters_update();
+
+  /**
+   * Calculate 'desired joint states' based on the 'measured force', 'reference joint state', and
+   * 'current_joint_state'.
+   *
+   * \param[in] current_joint_state
+   * \param[in] measured_wrench
+   * \param[in] reference_joint_state
+   * \param[in] period
+   * \param[out] desired_joint_state
+   */
+  controller_interface::return_type update(
+    const trajectory_msgs::msg::JointTrajectoryPoint & current_joint_state,
+    const geometry_msgs::msg::Wrench & measured_wrench,
+    const trajectory_msgs::msg::JointTrajectoryPoint & reference_joint_state,
+    const rclcpp::Duration & period,
+    trajectory_msgs::msg::JointTrajectoryPoint & desired_joint_states);
+
+  /**
+   * Set fields of `state_message` from current admittance controller state.
+   *
+   * \param[out] state_message
```

Comment:
Why not to add some description here to parameters? I understand that is “stating the obvious”, but it is also good to have some more information in the long term. Just write as you have to explain them to someone.

Reply:
done